### DISCUSSION TITLE
Is it possible to run FreeBSD binaries?

### DISCUSSION THREAD

**Discussion by megastallman** [jart/blink#167] [Ideas]
Hello community!
Blink becomes very interesting as a fully-fledged Linux emulator, though I admit that it is not intended to. Once I have even succeeded running Gimp, but QT5 software fails due to some Opengl problems...
Though I've got another idea. How much will it take us to run FreeBSD binaries on Blink? I know that:
- FreeBSD's ELFs start address is 200000, contrary to Linux's 100000.
- Syscall numbers are different.
- Kqueue needs to be emulated somehow, maybe through https://github.com/mheily/libkqueue
... and maybe a huge amount of other differences I'm not aware of.

**Comment by jart** [comment]
You would have to copy `blink/syscall.c` and rewrite a lot of it for FreeBSD. For example, all the magic numbers in that switch statement at the bottom of the file would need to change.

**Comment by megastallman** [comment]
It was another interesting experiment with AntigravityIDE, using all variants of Claude and Gemini AI. It has created a 1100 line patch https://github.com/jart/blink/commit/6979ce96a9465459c7cf6ffdc41eb7ac50b69eae for preliminary FreeBSD binary support. Both static and dynamic.
Though code quality is fairly poor. Complex applications like `md5sum`, `cc`, `w` or `sh` - just fail or hang forever, probably due to multithreaded execution failing at the moment. And it haven't yet reached the point where `libkqueue` is needed. So multithreading is the next challenge that AI can't figure out on its own yet.
```
$ file ../fbsd 
../fbsd: ELF 64-bit LSB executable, x86-64, version 1 (FreeBSD), statically linked, for FreeBSD 16.0 (1600003), FreeBSD-style, stripped
$ file ~/Software/blink/o/chroots/freebsd/usr/bin/uname
uname: ELF 64-bit LSB pie executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 16.0 (1600007), FreeBSD-style, stripped
$ ./o/blink/blink ../fbsd 
Hello World
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /usr/bin/uname -a
FreeBSD blink.local 16.0-RELEASE FreeBSD 16.0-RELEASE 6666666: Mon Jan 01 00:00:00 UTC 2024     root@blink.local:/usr/obj/usr/src/amd64.amd64/sys/GENERIC amd64
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/ls /
base.txz
bin
blink.log
boot
cat_debug.log
cat.log
COPYRIGHT
dev
echo.log
etc
final
hello.c
lib
libexec
media
mnt
net
proc
rescue
root
sbin
sys
tmp
usr
var
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/mkdir /qqq
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/mkdir /qqq
mkdir: /qqq: File exists
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/rmdir /qqq
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/rmdir /qqq
rmdir: /qqq: No such file or directory
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/cat /etc/resolv.conf
nameserver 8.8.8.8
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/echo eeeeeeeee
eeeeeeeee
$ ./o/blink/blink -C ~/Software/blink/o/chroots/freebsd /bin/hostname
blink.local
```

**Comment by megastallman** [comment]
Actually I've got even more success with running FreeBSD binaries. Here is my FREEBSD_AI branch: https://github.com/megastallman/blink/tree/FREEBSD_AI
Things that work: the most important terminal utilities, curl, wget, mc, nano, gcc, w, python3.11, pkg, even thunar runs with X11(Wayland is not yet fully working in FreeBSD itself). Also I've added root emulaton flag that allows pkg to unpack packages.

**Comment by megastallman** [comment]
Moved to my fork's master: https://github.com/megastallman/blink
